File: nodejs/src/rpc-client.js
import fetch from 'node-fetch';

export class RpcClient {
  /**
   * Send an RPC call
   * @param {string} url - The RPC server URL
   * @param {string} method - The RPC method to call
   * @param {Array} params - Parameters for the RPC method
   * @param {Object} [options] - Additional fetch options (e.g., auth)
   * @returns {Promise<any>} The result from the RPC server
   */
  async send(url, method, params = [], options = {}) {
    // Prepare JSON-RPC request
    const jsonReq = {
      jsonrpc: '2.0',
      id: Math.floor(Math.random() * 90000) + 10000,
      method,
      params,
    };

    // Prepare headers
    const headers = {
      'Content-Type': 'application/json',
    };

    // Add Basic Auth if auth option is provided
    if (options.auth) {
      const { username, password } = options.auth;
      const authString = Buffer.from(`${username}:${password}`).toString('base64');
      headers['Authorization'] = `Basic ${authString}`;
      console.log(`Generated Authorization header: Basic ${authString}`); // Debug
    }

    try {
      // Send HTTP request
      console.log(`Sending request to ${url} with headers:`, headers); // Debug
      const response = await fetch(url, {
        method: 'POST',
        headers,
        body: JSON.stringify(jsonReq),
      });

      if (!response.ok) {
        const body = await response.text();
        throw new Error(
          `Did not receive valid response from RPC server, got status ${response.status} with body: ${body}`
        );
      }

      const data = await response.json();
      console.log('Response:', data); // Debug

      if (data.error) {
        if (typeof data.error === 'object' && data.error.code && data.error.message) {
          throw new Error(`Received error from RPC server: Code (${data.error.code}) ${data.error.message}`);
        }
        throw new Error(`Received unexpected error from RPC server: ${JSON.stringify(data.error)}`);
      }

      if (!('result' in data)) {
        throw new Error(`Response missing 'result' field: ${JSON.stringify(data)}`);
      }

      return data.result;
    } catch (error) {
      console.error('Full error:', error); // Debug
      throw new Error(`Failed to send request to RPC server: ${error.message}`);
    }
  }
}

File: python/cicero_sdk/rpc_client.py

import json
import random
from typing import Any, Dict, List
import requests
from requests import Request, Response

class RpcClient:
    """RPC Client for making JSON-RPC 2.0 requests."""

    def send(self, url: str, method: str, params: List[Any] = None) -> Any:
        """
        Send an RPC call to the specified URL.

        Args:
            url (str): The RPC server URL.
            method (str): The RPC method to call.
            params (List[Any], optional): Parameters for the RPC method. Defaults to None.

        Returns:
            Any: The result from the RPC server.

        Raises:
            Exception: If the response is invalid or contains an error.
        """
        if params is None:
            params = []

        # Prepare JSON-RPC request
        json_req = {
            "jsonrpc": "2.0",
            "id": random.randint(10000, 99999),
            "method": method,
            "params": params
        }

        # Send HTTP request
        headers = {"Content-Type": "application/json"}
        try:
            response = requests.post(url, headers=headers, data=json.dumps(json_req))
        except requests.RequestException as e:
            raise Exception(f"Failed to send request to RPC server: {str(e)}")

        # Handle response
        if response.status_code != 200:
            raise Exception(
                f"Did not receive valid response from RPC server, got status {response.status_code} with body: {response.text}"
            )

        try:
            vars = response.json()
        except ValueError:
            raise Exception(f"Did not receive valid JSON object from RPC server, instead received: {response.text}")

        if "error" in vars and vars["error"] is not None:
            if isinstance(vars["error"], dict) and "code" in vars["error"] and "message" in vars["error"]:
                error_msg = f"Code ({vars['error']['code']}) {vars['error']['message']}"
            else:
                error_msg = f"Unexpected error format: {vars['error']}"
            raise Exception(f"Received error from RPC server: {error_msg}")

        if "result" not in vars:
            raise Exception(f"Response missing 'result' field: {vars}")

        return vars["result"]


